File: android/src/main/java/com/wix/RNCameraKit/torch/TorchModule.java
package com.wix.RNCameraKit.torch;

import android.content.Context;
import android.hardware.Camera;
import android.hardware.camera2.CameraManager;
import android.os.Build;

import com.facebook.react.bridge.Callback;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;

public class TorchModule extends ReactContextBaseJavaModule {
    private Boolean isTorchOn = false;
    private Camera camera;

    public TorchModule(ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @Override
    public String getName() {
        return "CKTorch";
    }

    public void switchState(Boolean newState, Callback successCallback, Callback failureCallback) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            try {
                CameraManager cameraManager = (CameraManager) getReactApplicationContext().getSystemService(Context.CAMERA_SERVICE);
                String cameraId;
                if (cameraManager != null) {
                    cameraId = cameraManager.getCameraIdList()[0];
                    cameraManager.setTorchMode(cameraId, newState);
                }
                successCallback.invoke(true);
            } catch (Exception e) {
                String errorMessage = e.getMessage();
                failureCallback.invoke("Error: " + errorMessage);
            }
        } else {
            Camera.Parameters params;

            if (!isTorchOn) {
                camera = Camera.open();
                params = camera.getParameters();
                params.setFlashMode(Camera.Parameters.FLASH_MODE_TORCH);
                camera.setParameters(params);
                camera.startPreview();
                isTorchOn = true;
            } else {
                params = camera.getParameters();
                params.setFlashMode(Camera.Parameters.FLASH_MODE_OFF);

                camera.setParameters(params);
                camera.stopPreview();
                camera.release();
                isTorchOn = false;
            }
        }
    }
}

File: android/src/main/java/com/wix/RNCameraKit/torch/TorchModule.kt
package com.wix.RNCameraKit.torch

import android.content.Context
import android.hardware.Camera
import android.hardware.camera2.CameraManager
import android.os.Build
import com.facebook.react.bridge.Callback
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod

class TorchModule(private val myReactContext: ReactApplicationContext)
    : ReactContextBaseJavaModule(myReactContext) {
    private var isTorchOn: Boolean? = false
    private var camera: Camera? = null

    override fun getName(): String {
        return "CKTorch"
    }

    @ReactMethod
    fun switchState(newState: Boolean?, successCallback: Callback, failureCallback: Callback) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            val cameraManager = this.myReactContext.getSystemService(Context.CAMERA_SERVICE) as CameraManager
            try {
                val cameraId = cameraManager.cameraIdList[0]
                cameraManager.setTorchMode(cameraId, newState!!)
                successCallback.invoke(true)
            } catch (e: Exception) {
                val errorMessage = e.message
                failureCallback.invoke("Error: $errorMessage")
            }

        } else {
            val params: Camera.Parameters

            if (!isTorchOn!!) {
                camera = Camera.open()
                params = camera!!.parameters
                params.flashMode = Camera.Parameters.FLASH_MODE_TORCH
                camera!!.parameters = params
                camera!!.startPreview()
                isTorchOn = true
            } else {
                params = camera!!.parameters
                params.flashMode = Camera.Parameters.FLASH_MODE_OFF

                camera!!.parameters = params
                camera!!.stopPreview()
                camera!!.release()
                isTorchOn = false
            }
        }
    }
}